consensus/enclave/impl/src/constant_time_token_map.rs

```diff
@@ -0,0 +1,331 @@
+    /// that we are constant-time with respect to which token id it is that
+    /// is in the map.
+    pub fn get(&self, token_id: &TokenId) -> Option<T> {
+        let mut success = Choice::from(0);
+        let mut result = T::default();
+        for (key, val) in self.storage.iter() {
+            let found = key.ct_eq(token_id);
+            result.conditional_assign(val, found);
+            success.conditional_assign(&found, found);
+        }
+        if success.unwrap_u8() == 0 {
+            None
+        } else {
+            Some(result)
+        }
+    }
+
+    /// Set the value associated to a token id in constant time.
+    ///
+    /// This does not insert the token id if it was not set at initialization.
+    /// If the token id is not present in the map, this function has no
+    /// side-effect and returns false.
+    // Note: this is not used right now, but we are leaving it in case we have
+    // dynamic minimum fees at some point in the future.
+    #[allow(dead_code)]
+    pub fn set(&mut self, token_id: &TokenId, new_val: T) -> Choice {
+        let mut success = Choice::from(0);
+        for (key, val) in self.storage.iter_mut() {
+            let found = key.ct_eq(token_id);
+            val.conditional_assign(&new_val, found);
+            success.conditional_assign(&found, found);
+        }
+        success
+    }
+
+    /// Add the value associated to a token id in constant time.
+    ///
+    /// This does not add the token id if it was not set at initialization.
+    /// If the token id is not present in the map, this function has no
+    /// side-effect and returns false.
+    pub fn add(&mut self, token_id: &TokenId, addend: impl Copy + Add<T, Output = T>) -> Choice {
+        let mut success = Choice::from(0);
+        for (key, val) in self.storage.iter_mut() {
+            let found = key.ct_eq(token_id);
+            val.conditional_assign(&(addend + *val), found);
+            success.conditional_assign(&found, found);
+        }
+        success
+    }
+
+    /// Get the number of key-value pairs in the map.
+    pub fn len(&self) -> usize {
+        self.storage.len()
+    }
+
+    /// Iterate over all entries in the map.
+    pub fn iter(&self) -> impl Iterator<Item = &(TokenId, T)> {
+        self.storage.iter()
+    }
+
+    /// Get an iterator over the keys of this map.
+    pub fn keys(&self) -> impl Iterator<Item = &TokenId> {
+        self.storage.iter().map(|(key, _val)| key)
+    }
+}
+
+impl<T: ConstantTimeEq + ConditionallySelectable + Default> CtTokenMap<T> {
+    /// Sort any zero values to the end in constant time.
+    ///
+    /// For accumulated fees, we want to create as few fee outputs as we can get
+    /// away with, but without revealing any information about if some
+    /// transactions had the same token id.
+    ///
+    /// To do this, we want to be able to collect all the "zero" values at the
+    /// end of the buffer.
```

Comment:
Naive ask: Why?

Reply:
it supports this part:

https://github.com/mobilecoinfoundation/mobilecoin/pull/1780/files/650b010c47f29ed892497f6f12a4aaa8577a8a3e#r846287630